Userland/Libraries/LibGfx/QOILoader.cpp

```diff
@@ -0,0 +1,285 @@
+/*
+ * Copyright (c) 2021, Linus Groh <linusg@serenityos.org>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include <AK/Endian.h>
+#include <AK/MemoryStream.h>
+#include <LibGfx/Bitmap.h>
+#include <LibGfx/QOILoader.h>
+
+namespace Gfx {
+
+static constexpr auto QOI_MAGIC = "qoif"sv;
+static constexpr u8 QOI_OP_RGB = 0b11111110;
+static constexpr u8 QOI_OP_RGBA = 0b11111111;
+static constexpr u8 QOI_OP_INDEX = 0b00000000;
+static constexpr u8 QOI_OP_DIFF = 0b01000000;
+static constexpr u8 QOI_OP_LUMA = 0b10000000;
+static constexpr u8 QOI_OP_RUN = 0b11000000;
+static constexpr u8 QOI_MASK_2 = 0b11000000;
+static constexpr u8 END_MARKER[] = { 0, 0, 0, 0, 0, 0, 0, 1 };
+
+static ErrorOr<QOIHeader> decode_qoi_header(InputMemoryStream& stream)
+{
+    QOIHeader header;
+    stream >> Bytes { &header, 14 }; // sizeof includes two bytes of padding!
+    if (stream.handle_any_error())
+        return Error::from_string_literal("QOIImageDecoderPlugin: Invalid QOI image: end of stream while reading header"sv);
+    if (StringView { header.magic, sizeof(header.magic) } != QOI_MAGIC)
+        return Error::from_string_literal("QOIImageDecoderPlugin: Invalid QOI image: incorrect header magic"sv);
+    header.width = AK::convert_between_host_and_big_endian(header.width);
+    header.height = AK::convert_between_host_and_big_endian(header.height);
+    return header;
+}
+
+static ErrorOr<Color> decode_qoi_op_rgb(InputMemoryStream& stream)
+{
+    u8 bytes[4] {};
```

Comment:
Zero initialising these arrays seems like a waste, we just immediately overwrite the zeros...

Reply:
Agreed